### ISSUE TITLE
new c lib: add method to statically define im2 vector table for z80 and z180 targets

### ISSUE THREAD

**Issue by aralbrec** [z88dk/z88dk#76]
Currently the library provides a method to create an im2 vector table programmatically via functions in <z80.h>

The user program is expected to create the table at a fixed available location in memory, perhaps initializing with suitable **memset()** or **memcpy()** call.  Into that table, isr addresses can optionally be written using **im2_install_isr()**.  The library also provides a **generic isr** that is capable of collecting a list of functions that is executed on interrupt to a specific vector.  Once the table is initialized, the program is expected to call **im2_init(vector table address)** to load register **I** with the vector table address and enable interrupts to get things going. 

The table can also be created in a separate asm file and this can be copied to its final location in memory at startup using suitable asm in **section code_crt_init** or a **memcpy()** in main().

Having to do these things increases program size a small amount and is less convenient than necessary especially for the embedded z80 and z180 targets.

The change being proposed is motivated primarily by the embedded z180 target but this could also be applied to the embedded z80 target.  Most z80 computers do not use the im2 mode so other z80 targets would likely prefer that such a thing is not enabled by default.

The z180 must always have an im2 vector table.  It does have a z80 mode on interrupt pin /INT0 that can operate in the usual im0, im1, im2 modes but its other interrupts (/INT1, /INT2, internal devices) can only work in im2 mode.  So for the z180, you will always want im2 enabled unless you do not use the built in z180 peripherals and only use the z180 as a fast z80.  In that case, I would say the embedded z80 target should be used rather than the embedded z180.

The best place for the z180's interrupt vector table is in the page zero crt code that sets up the restarts.  There the im2 vector table can be located on a 256-byte boundary with the page zero include padding using common library routines so that any necessary padding to advance to the correct alignment is useful.  The other advantage is that any page zero code will be in the z180's common segment rather than in a banked segment where logical address may not be the same as physical address.  The vector table **must** be located in physical space and that can only be done in the common segment.  Similarly, the interrupt service routines themselves need to be in the common segment and for that the recommendation is going to be to assign ISR code into the already existing **code_crt_common** section.

The leading method for implementing the vector table is currently the addition of a **code_vector_table** section that will be aligned to a 256-byte boundary inside the crt when a suitable pragma is made active in the compile.  The program can then assign vector addresses to that section in any way deemed appropriate.

**Comment by aralbrec** [comment]
For all targets in the new c library a new pragma has been defined "CRT_GENERATE_VECTOR_TABLE" that, if set non-zero, causes the page zero code generated by z88dk at address 0 to include space for an interrupt vector table.  This table has org 0x100 and has a section "code_vector" defined for it.

For all z80 targets CRT_GENERATE_VECTOR_TABLE=0 and for all z180 targets CRT_GENERATE_VECTOR_TABLE=1 by default.

The programmer can add vectors to the table simply assigning isr address to the section.  For example, a file "interrupt.asm" might be part of a project that maintains the vector table:

```
SECTION code_vector

interrupt_vector_table:

   ; vector 0
   EXTERN isr_clock
   defw isr_clock

   ; vector 2
   EXTERN isr_pio_0
   defw isr_pio_0
```

Then the table would become part of the binary simply by adding "interrupt.asm" to the compile line.

Note that space for the vector table is contingent on z88dk's page zero code being used.  This happens if CRT_ORG_CODE=0 and the user hasn't defined CRT_INCLUDE_PREAMBLE which inserts code right at the beginning of the binary so that the org is pushed off address 0.  In cases where z88dk's page zero is not used, any assignment to "section code_vector" will result in a separately generated binary "foo_code_vector.bin" representing the vector table that is not part of the output binary.

**Comment by aralbrec** [comment]
Some information about the im2 vector table has been added to the wiki:
https://www.z88dk.org/wiki/doku.php?id=libnew:target_embedded#interrupt_service_routines
